drivers/apc_modbus.c

```diff
@@ -0,0 +1,1224 @@
+			status_set("BOOST");
+		}
+		if (value & (1 << 6)) {
+			status_set("TRIM");
+		}
+
+		_apc_modbus_process_registers(apc_modbus_register_map_dynamic, regbuf, 32, 128);
+	} else {
+		dstate_datastale();
+		return;
+	}
+
+	/* Static Data */
+	if (_apc_modbus_read_registers(modbus_ctx, 1026, 7, regbuf)) {
+		_apc_modbus_process_registers(apc_modbus_register_map_static, regbuf, 7, 1026);
+	} else {
+		dstate_datastale();
+		return;
+	}
+
+	alarm_commit();
+	status_commit();
+	dstate_dataok();
+}
+
+void upsdrv_shutdown(void)
+{
+	/* TODO: replace with a proper shutdown function */
+	upslogx(LOG_ERR, "shutdown not supported");
+	set_exit_flag(-1);
+}
+
+void upsdrv_help(void)
+{
+}
+
+void upsdrv_makevartable(void)
+{
+# if defined NUT_MODBUS_HAS_USB
+	nut_usb_addvars();
+# endif /* defined NUT_MODBUS_HAS_USB */
+
+# if defined NUT_MODBUS_HAS_USB
+	addvar(VAR_VALUE, "porttype", "Modbus port type (serial, tcp, usb, default=usb)");
+# else
+	addvar(VAR_VALUE, "porttype", "Modbus port type (serial, tcp, default=serial)");
+# endif /* defined NUT_MODBUS_HAS_USB */
+	addvar(VAR_VALUE, "slaveid", "Modbus slave id (default=1)");
+
+	/* Serial RTU parameters */
+	addvar(VAR_VALUE, "baudrate", "Modbus serial RTU communication speed in baud (default=9600)");
+	addvar(VAR_VALUE, "parity", "Modbus serial RTU parity (N=none, E=even, O=odd, default=N)");
+	addvar(VAR_VALUE, "databits", "Modbus serial RTU data bit count (default=8)");
+	addvar(VAR_VALUE, "stopbits", "Modbus serial RTU stop bit count (default=1)");
+}
+
+# if defined NUT_MODBUS_HAS_USB
+static int _apc_modbus_usb_device_match_func(USBDevice_t *hd, void *privdata)
+{
+	NUT_UNUSED_VARIABLE(privdata);
+
+	if (hd->VendorID != 0x051d || hd->ProductID != 0x0003) {
+		return 0;
+	}
+
+	return 1;
+}
+
+static USBDeviceMatcher_t apc_modbus_usb_device_matcher = { &_apc_modbus_usb_device_match_func, NULL, NULL };
+
+static void _apc_modbus_usb_lib_to_nut(const modbus_usb_device_t *device, USBDevice_t *out)
+{
+	/* This makes a USBDevice_t from modbus_usb_device_t so we can use our matchers */
+
+	static char bus_buf[4], device_buf[4], bus_port_buf[4];
+
+	out->VendorID = device->vid;
+	out->ProductID = device->pid;
+	out->Vendor = device->vendor_str;
+	out->Product = device->product_str;
+	out->Serial = device->serial_str;
+	out->bcdDevice = device->bcd_device;
+
+	snprintf(bus_buf, sizeof(bus_buf), "%03d", device->bus);
+	out->Bus = bus_buf;
+
+	snprintf(device_buf, sizeof(device_buf), "%03d", device->device_address);
+	out->Device = device_buf;
+
+#if (defined WITH_USB_BUSPORT) && (WITH_USB_BUSPORT)
+	snprintf(bus_port_buf, sizeof(bus_port_buf), "%03d", device->bus_port);
+	out->BusPort = bus_port_buf;
+#endif
+}
+
+static int _apc_modbus_usb_callback(const modbus_usb_device_t *device)
+{
+	HIDDesc_t *hid_desc;
+	size_t i;
+	HIDData_t *hid_cur_item, *hid_rtu_rx = NULL, *hid_rtu_tx = NULL;
+	HIDNode_t hid_cur_usage;
+	USBDeviceMatcher_t *current_matcher;
+
+	if (device->hid_report_descriptor_buf == NULL) {
```

Comment:
I think the report descriptor check should be later in the function, there are reasons why the report descriptor can not be read but the device might still be usable. We should just rely on the default report ids.